File: go/day02/day02.go
package day02

import (
	"aoc23/inputs"
	"fmt"
	"strconv"
	"strings"
)

func Run() [2]int {
	games := parse(inputs.Input02)
	p1 := part1(games)
	p2 := part2(games)
	return [2]int{p1, p2}
}

const (
	thresRed   = 12
	thresGreen = 13
	thresBlue  = 14
)

type game struct {
	num    int
	result [][3]int
}

func parse(input string) []game {
	lines := strings.Split(input, "\n")
	lines = lines[:len(lines)-1]
	games := make([]game, len(lines))

	for i, line := range lines {
		games[i] = parseLine(line)
	}

	return games
}

func parseNum(s string) (int, int) {
	acc := ""
	for _, c := range s {
		if c >= '0' || c <= '9' {
			acc += string(c)
		}
	}
	n, _ := strconv.Atoi(acc)
	return n, len(acc)
}

func parseLine(line string) game {
	var g game
	fmt.Sscanf(line, "Game %d:", &g.num)
	line = strings.Split(line, ":")[1]
	parts := strings.Split(line, ";")
	g.result = make([][3]int, len(parts))
	for i, p := range parts {
		for _, c := range strings.Split(p, ",") {
			var num int
			fmt.Sscanf(c, "%d", &num)
			if strings.HasSuffix(c, "red") {
				g.result[i][0] = num
			} else if strings.HasSuffix(c, "green") {
				g.result[i][1] = num
			} else if strings.HasSuffix(c, "blue") {
				g.result[i][2] = num
			} else {
				panic(fmt.Sprintf("unknown color: \"%s\"", c))
			}
		}
	}
	return g
}

func part2(games []game) int {
	sum := 0
	for _, g := range games {
		maxs := [3]int{0, 0, 0}
		for _, r := range g.result {
			if r[0] > maxs[0] {
				maxs[0] = r[0]
			}
			if r[1] > maxs[1] {
				maxs[1] = r[1]
			}
			if r[2] > maxs[2] {
				maxs[2] = r[2]
			}
		}
		sum += maxs[0] * maxs[1] * maxs[2]
	}
	return sum
}

func part1(games []game) int {
	sum := 0
	for _, g := range games {
		possible := true
		for _, r := range g.result {
			if r[0] > thresRed || r[1] > thresGreen || r[2] > thresBlue {
				possible = false
			}
		}
		if possible {
			sum += g.num
		}
	}
	return sum
}

File: rs/src/day02.rs
pub fn run() -> [u32; 2] {
    let input = include_str!("../inputs/input02.txt");
    let parsed = parse(input);
    let p1 = part01(&parsed);
    let p2 = part02(&parsed);
    [p1, p2]
}

const THRES: [u32; 3] = [12, 13, 14];

fn part01(input: &[Game]) -> u32 {
    input
        .iter()
        .filter(|g| {
            g.results
                .iter()
                .all(|r| r.iter().zip(THRES.iter()).all(|(x, th)| x <= th))
        })
        .map(|g| g.num)
        .sum()
}

fn part02(input: &[Game]) -> u32 {
    input
        .iter()
        .map(|g| {
            g.results.iter().fold([0, 0, 0], |mut acc, r| {
                (0..3usize).for_each(|i| {
                    if r[i] > acc[i] {
                        acc[i] = r[i];
                    }
                });
                acc
            })
        })
        .map(|r| r.iter().product::<u32>())
        .sum()
}

#[derive(Debug)]
struct Game {
    num: u32,
    results: Vec<[u32; 3]>,
}

fn parse(input: &str) -> Vec<Game> {
    input
        .lines()
        .map(parse_line)
        .into_iter()
        .enumerate()
        .map(|(i, mut g)| {
            g.num = u32::try_from(i).unwrap() + 1u32;
            g
        })
        .collect()
}

fn parse_line(line: &str) -> Game {
    Game {
        num: 0,
        results: line
            .split(':')
            .last()
            .unwrap()
            .split(';')
            .map(parse_result)
            .collect(),
    }
}

fn parse_result(line: &str) -> [u32; 3] {
    let mut result = [0, 0, 0];

    line.split(',').for_each(|x| {
        let digit = parse_one_digit(x);
        if x.ends_with("red") {
            result[0] = digit;
        } else if x.ends_with("green") {
            result[1] = digit;
        } else if x.ends_with("blue") {
            result[2] = digit;
        }
    });

    result
}

fn parse_one_digit(line: &str) -> u32 {
    line.split_whitespace()
        .next()
        .unwrap()
        .parse::<u32>()
        .unwrap()
}
